pypit/core/artracewave.py

```diff
@@ -0,0 +1,1235 @@
+    """
+    sz_x, sz_o = tilts.shape
+    dszx = (sz_x-1.0)
+
+    tiltsimg = np.zeros((sz_x,sz_y), dtype=float)
+    for o in range(sz_o):
+        for x in range(sz_x):
+            ow = (rordloc[x,o]-lordloc[x,o])/2.0
+            oc = (rordloc[x,o]+lordloc[x,o])/2.0
+            ymin = int(oc-ow) - pad
+            ymax = int(oc+ow) + 1 + pad
+            # Check we are in bounds
+            if ymin < 0:
+                ymin = 0
+            elif ymax < 0:
+                continue
+            if ymax > sz_y-1:
+                ymax = sz_y-1
+            elif ymin > sz_y-1:
+                continue
+            # Set the tilt value at each pixel in this row
+            for y in range(ymin, ymax):
+                yv = (y-lordloc[x, o])/ow - 1.0
+                tiltsimg[x,y] = (tilts[x,o]*yv + x)/dszx
+    return tiltsimg
+
+
+def trace_tilt(ordcen, rordloc, lordloc, det, msarc, slitnum, settings_spect,
+               tilt_settings, censpec=None, maskval=-999999.9,
+               trthrsh=1000.0, nsmth=0, method="fweight", wv_calib=None):
+    """
+    This function performs a PCA analysis on the arc tilts for a single spectrum (or order)
+               trthrsh=1000.0, nsmth=0):
+
+    Parameters
+    ----------
+    slf
+    det
+    msarc
+    slitnum : int
+      Slit number, here indexed from 0
+    censpec
+    maskval
+    trthrsh
+    nsmth
+    method : str (fweight or cc)
+
+    Returns
+    -------
+    trcdict : dict
+
+    """
+    #if settings.argflag['trace']['slits']['tilts']['idsonly']:
+    def pad_dict(indict):
+        """ If an arc line is considered bad, fill the
+        dictionary arrays with null values
+        """
+        indict["xtfit"].append(None)
+        indict["ytfit"].append(None)
+        indict["wmask"].append(None)
+        return indict
+
+    # from pypit import arcyutils
+    dnum = arparse.get_dnum(det)
+
+    msgs.work("Detecting lines for slit {0:d}".format(slitnum+1))
+    tampl, tcent, twid, w, _ = ararc.detect_lines(censpec)
+    satval = settings_spect['saturation']*settings_spect['nonlinear']
+    # Order of the polynomials to be used when fitting the tilts.
+    arcdet = (tcent[w]+0.5).astype(np.int)
+    ampl = tampl[w]
+
+    # Determine the best lines to use to trace the tilts
+    ncont = 15
+    aduse = np.zeros(arcdet.size, dtype=np.bool)  # Which lines should be used to trace the tilts
+    w = np.where(ampl >= trthrsh)
+    aduse[w] = 1
+    # Remove lines that are within ncont pixels
+    nuse = np.sum(aduse)
+    detuse = arcdet[aduse]
+    idxuse = np.arange(arcdet.size)[aduse]
+    olduse = aduse.copy()
+    for s in range(nuse):
+        w = np.where((np.abs(arcdet-detuse[s]) <= ncont) & (np.abs(arcdet-detuse[s]) >= 1.0))[0]
+        for u in range(w.size):
+            if ampl[w[u]] > ampl[olduse][s]:
+                aduse[idxuse[s]] = False
+                break
+    # Restricted to ID lines? [introduced to avoid LRIS ghosts]
+    if tilt_settings['tilts']['idsonly']:
```

Comment:
Now that I think about it, by using `idsonly`, it may allow us to remove the code to generate an arc saturation mask, which was used in the echelle code to mask saturation streaks from arc lines in neighbouring orders. Maybe add a todo item at the relevant point in the code where `gen_satmask` is called?